tour.md

```diff
@@ -1,730 +1,493 @@
+
+fn name[T, E](a: A, b: B, c: C) -> T
+  where
+    A: IA,
+    B: IB,
+    C: IC
+{
+  /*...*/
+}
+
+// anonymous function
+square := \x {x*x}
+
+fn fib(n: int): int {
+  if n < 2 { n }
+  else n * fib(n - 1)
+}
+
+// return may be used to return early
+fn test() {
+  if something() { return }
+  print("yo")
+}
+
+fn split(s: string, sep: string) -> [string] {
+  out := [];
+  start := 0;
+  end := 0;
+
+  for ch in s {
+    if ch == sep {
+      out.push(s[start..end])
+      start = end;
+    } else {
+      end += 1;
+    }
+  }
+
+  if current.len > 0 {
+    out.push(current);
+  }
+
+  out
+}
+
+print(split("a,b,c", ","))
+```
+
+Classes, traits
+
+```rust
+trait Default {
+  fn default() -> Self;
+}
+
+trait Iterate {
+  type Iter: Iterator;
+  fn iter(self) -> Iter;
+}
+
+trait Iterator {
+  type Item;
+  fn next(self) -> Item?;
+}
+
+class Node<T> {
+  value: T;
+  next: Node[T]?;
+}
+
+class List[T] {
+  @head: Node[T]?;
+
+  fn new() -> Self {
+    List(head: null);
+  }
+
+  fn prepend(self, value: T) {
+    next := self.head;
+    self.head = Node(value);
+    self.head.next = next;
+  }
+
+  impl Iterate {
+    type Iter = ListIter[T];
+    fn iter(self) -> Iter {
+      ListIter(node: self.head)
+    }
+  }
+}
+
+class ListIter[T] {
+  node: Node[T];
+
+  impl Iterator {
+    type Item = Node[T];
+    fn next(self) -> Item? {
+      node := self.node;
+      self.node = node.next;
+      node.value
+    }
+  }
+}
+
+list := List.new();
+
+for i in 0..10 {
+  list.prepend(i);
+}
+
+for val in list {
+  print(val);
+}
+```
+
+Exceptions
+
+```rust
+class A {}
+class B {}
+class C {}
+
+fn f() {
+  throw A()
+}
+
+v := try f() {
+  A => ...,
+  B => ...,
+  C => ...,
+}
+
+try {
+  a();
+  b();
+  c();
+} catch {
+  A => ...
+  B => ...
+  C => ...
+}
+```
+
+Modules
+
+```rust
+import module
+import module.A
+import module.{A, B}
+
+import nested.module.A
+import nested.module.{A, B}
+
+import {
+  a.{A, B},
+  b.{C, D},
+  nested.{
+    d.{E, F},
+    f.{G, H},
+  }
+}
+
+// you can export functions, classes, enums, and types
+export fn a() {}
+export class B {}
+export enum C {}
+export type D {}
+
+// exports may also be grouped into a block export
+fn e() {}
+traits F
+class G {}
+// symbols exported this way may be renamed using `as`
+export { e as g, F as E, G as F }
+
+// module imports may be arbitrarily re-ordered by the runtime, which means that
+// there is no guarantee of the order in which the "side effects" of imported modules are processed.
+// the only guarantee is that code at the top-level in imported modules will be executed exactly once.
+```
+
+Concurrency
+
+```rust
+/*
+Everything is asynchronous by default.
+All execution happens within the context of a task.
+Whenever you do any I/O, the task is blocked until that I/O finishes,
+but other tasks may be scheduled to run during during that time.
+
+you may also know tasks under names such as:
+- stackful coroutines
+- goroutines

```

Comment:
Goroutines may yield at any point IIRC (preemptive multitasking), not only well-defined await points (cooperative multistaking), like in our case. But I it may be a good idea to reserve the option to make the runtime preemptive by not offering this guarantee.

Reply:
True, I'll add that in